### PR TITLE
Rewrite to Call NNSymbolSetAttrs

### PR DESCRIPTION
**Pull request by toddsundsted**
Addresses issue https://github.com/mrkn/mxnet.rb/issues/40.

### CODE REVIEW

File: ext/mxnet/mxnet_internal.h [hunk 1]
```diff
@@ -230,9 +230,10 @@ struct mxnet_api_table {
                           const char *key,
                           const char **out,
                           int *success);
-  int (* MXSymbolSetAttr)(SymbolHandle symbol,
-                          const char *key,
-                          const char *value);
+  int (* NNSymbolSetAttrs)(SymbolHandle symbol,
+                           mx_uint num_attrs,
```
Comment (mrkn) [inline ext/mxnet/mxnet_internal.h]: I want to add and use `nn_uint` like `mx_uint`.

File: ext/mxnet/symbol.c [hunk 2]
```diff
@@ -280,29 +287,57 @@ symbol_set_attributes_i(VALUE key, VALUE value, VALUE arg)
...
  */
 static VALUE
 symbol_set_attributes(int argc, VALUE *argv, VALUE obj)
 {
-  VALUE kwargs;
   SymbolHandle handle;
+  struct set_attributes_args args;
+  VALUE kwargs, gc_guard = Qnil, keys_str = Qnil, vals_str = Qnil;
+  char const **keys = NULL, **vals = NULL;
+  int num_attrs = 0;
```
Comment (mrkn) [inline ext/mxnet/symbol.c]: It should be `nn_uint` (or `mx_uint`) instead of `int`.

File: ext/mxnet/symbol.c [hunk 3]
```diff
@@ -280,29 +287,57 @@ symbol_set_attributes_i(VALUE key, VALUE value, VALUE arg)
...
 
   rb_scan_args(argc, argv, ":", &kwargs);
 
   if (!NIL_P(kwargs)) {
+    if (RHASH_SIZE(kwargs) > INT_MAX) {
```
Comment (mrkn) [inline ext/mxnet/symbol.c]: `UINT_MAX`

File: ext/mxnet/symbol.c [hunk 4]
```diff
@@ -280,29 +287,57 @@ symbol_set_attributes_i(VALUE key, VALUE value, VALUE arg)
...
+      rb_raise(rb_eArgError, "too many attributes");
+    }
+    num_attrs = (int)RHASH_SIZE(kwargs);
```
Comment (mrkn) [inline ext/mxnet/symbol.c]: Same as above.

File: ext/mxnet/symbol.c [hunk 5]
```diff
@@ -280,29 +287,57 @@ symbol_set_attributes_i(VALUE key, VALUE value, VALUE arg)
...
+
+    gc_guard = rb_ary_new_capa((long)num_attrs * 2);
+    keys_str = rb_str_tmp_new(sizeof(char const *) * num_attrs);
+    keys = (char const**)RSTRING_PTR(keys_str);
```
Comment (mrkn) [inline ext/mxnet/symbol.c]: Please put a single white space between `char const` and `**`.

File: ext/mxnet/symbol.c [hunk 6]
```diff
@@ -280,29 +287,57 @@ symbol_set_attributes_i(VALUE key, VALUE value, VALUE arg)
...
+    vals_str = rb_str_tmp_new(sizeof(char const *) * num_attrs);
+    vals = (char const**)RSTRING_PTR(vals_str);
```
Comment (mrkn) [inline ext/mxnet/symbol.c]: Same as above.

File: ext/mxnet/mxnet_internal.h (hunk 1 above)
Comment (toddsundsted) [reply ext/mxnet/mxnet_internal.h]: fixed in c5227f7

File: ext/mxnet/symbol.c (hunk 2 above)
Comment (toddsundsted) [reply ext/mxnet/symbol.c]: fixed in c5227f7

File: ext/mxnet/symbol.c (hunk 4 above)
Comment (toddsundsted) [reply ext/mxnet/symbol.c]: fixed in c5227f7

File: ext/mxnet/symbol.c (hunk 5 above)
Comment (toddsundsted) [reply ext/mxnet/symbol.c]: fixed in b97616b

File: ext/mxnet/symbol.c (hunk 6 above)
Comment (toddsundsted) [reply ext/mxnet/symbol.c]: fixed in b97616b

File: ext/mxnet/symbol.c (hunk 3 above)
Comment (toddsundsted) [reply ext/mxnet/symbol.c]: fixed in b97616b

**Comment by mrkn** [comment]
@toddsundsted Thank you very much!

**Pull request merged**